src/aspire/abinitio/commonline_sync3n.py

```diff
@@ -0,0 +1,1025 @@
+            """
+            # Get inistial estimate for Pij
+            P, sigma, Pij, hist = self._triangle_scores(Rijs, hist, Pmin, Pmax)
+
+            # Check if P and Pij are consistent
+            mean_Pij = np.mean(Pij)
+            too_low = P < mean_Pij / permitted_inconsistency
+            too_high = P > mean_Pij * permitted_inconsistency
+            inconsistent = too_low | too_high
+
+            # Check trend
+            if prev_too_low is not None and too_low != prev_too_low:
+                p_domain_limit = np.sqrt(p_domain_limit)
+
+            # define limits for next P estimation
+            if too_high:
+                if P < min_p_permitted:
+                    logger.error(
+                        "Triangles Scores are poorly distributed, whatever small P we force."
+                    )
+
+                if Pmax is not None:
+                    Pmax = Pmax * p_domain_limit
+                else:
+                    Pmax = P
+
+                Pmin = Pmax * p_domain_limit
+            else:  # too low
+                if Pmin is not None:
+                    Pmin = Pmin / p_domain_limit
+                else:
+                    Pmin = P
+
+                Pmax = Pmin / p_domain_limit
+
+            return inconsistent, Pij, (too_low, Pmin, Pmax, hist)
+
+        # Repeat iteratively until estimations of P & Pij are consistent
+        i = 0
+        res = (None,) * 4
+        inconsistent = True
+        while inconsistent and i < max_iterations:
+            inconsistent, Pij, res = _body(*res)
+            i += 1
+
+        # Pack W
+        W = np.zeros((self.n_img, self.n_img))
+        idx = 0
+        for i in range(self.n_img):
+            for j in range(i + 1, self.n_img):
+                W[i, j] = Pij[idx]
+                W[j, i] = Pij[idx]
+                idx += 1
+
+        return W
+
+    def _triangle_scores_inner(self, Rijs):
+        """
+        Computes histogram of `triangle scores`.
+
+        Wrapper for cpu/gpu dispatch.
+
+        :param Rijs: nchoose2 by 3 by 3 array of rotations.
+        :return: Histogram of triangle scores.
+        """
+
+        # host/gpu dispatch
+        if self._gpu_module:
+            scores_hist = self._triangle_scores_inner_cupy(Rijs)
+        else:
+            scores_hist = self._triangle_scores_inner_host(Rijs)
+
+        return scores_hist
+
+    def _triangle_scores_inner_host(self, Rijs):
+        """
+        See _triangle_scores_inner.
+
+        CPU implementation.
+        """
+
+        # The following is adopted from Matlab triangle_scores_mex.c
+
+        # Initialize probability result arrays
+        scores_hist = np.zeros(self.hist_intervals, dtype=np.uint32)
+        h = 1 / self.hist_intervals
+
+        c = np.empty((4), dtype=Rijs.dtype)
+        for i in trange(self.n_img, desc="Computing triangle scores"):
+            for j in range(
+                i + 1, self.n_img - 1
+            ):  # check bound (taken from MATLAB mex)
```

Comment:
I'm not sure what you mean here, because your suggection makes the same sequence.

Is it that you just dont like the two empty loops reulting from the outer `range(n_img)` vs `range(n_img-2)`? I'm assuming that's it and updated, but if it was something else lmk.

```
>>> n_img = 5
>>> for i in range(n_img - 2):
...   for j in range(i + 1, n_img-1):
...      for k in range(j + 1, n_img):
...        print(i,j,k)
... 
0 1 2
0 1 3
0 1 4
0 2 3
0 2 4
0 3 4
1 2 3
1 2 4
1 3 4
2 3 4
>>> for i in range(n_img):
...   for j in range(i + 1, n_img-1):
...      for k in range(j + 1, n_img):
...        print(i,j,k)
... 
0 1 2
0 1 3
0 1 4
0 2 3
0 2 4
0 3 4
1 2 3
1 2 4
1 3 4
2 3 4
```

Reply:
There was a comment about checking the bound. I was just saying that the bound is good and giving a suggestion if you wanted to make the bounds explicit (ie. no empty loops).

Reply:
Okay cool.  Yeah, I didn't like those bounds when I ported them, but maybe they were the secret sauce 🤷‍♂️  lol.  Thanks for double checking.